### PR TITLE
Validation Refactor: GE and Additional Rules

### PR DESCRIPTION
**Pull request by GiaJordan**
First round of the validation refactor, includes hybrid validation approach utilizing great expectations and cross manifest validation in its first iteration

### CODE REVIEW

File: great_expectations/expectations/Manifest_test_suite.json [hunk 1]
```diff
@@ -0,0 +1,77 @@
+{
```
Comment (milen-sage) [inline great_expectations/expectations/Manifest_test_suite.json:1]: We don't need to track this file.

File: poetry.lock [hunk 2]
```diff
@@ -1,3 +1,22 @@
+[[package]]
```
Comment (milen-sage) [inline poetry.lock:1]: don't track poetry.lock the toml should be good enough

File: schematic/models/validate_attribute.py [hunk 3]
```diff
@@ -421,3 +505,66 @@ def url_validation(self, val_rule: str, manifest_col: str) -> List[List[str]]:
...
+        #parse sources and targets
+        [source_component, source_attribute] = val_rule.split(" ")[1].split(".")
+        [target_component, target_attribute] = val_rule.split(" ")[2].split(".")
+
+        #Get IDs of manifests with target component
+        syn, target_IDs=ValidateAttribute.get_target_manifests(target_component)
+
+        #Read each manifest
+        for target_manifest_ID in target_IDs:
+            entity = syn.get(target_manifest_ID)
```
Comment (milen-sage) [inline schematic/models/validate_attribute.py:526]: Let's add syn.get wrapper in SynapseStore. We can name it getAsset() and pass the manifest ID. That way we can decouple a bit from Synapse in this validation module.

**Review by milen-sage — CHANGES_REQUESTED** [review]
Some comments in code

File: poetry.lock (hunk 2 above)
Comment (GiaJordan) [reply poetry.lock]: Reverting the `.lock` file to its previous version and only changing the `.toml` seems to be causing errors during library installation

**Comment by rrchai** [comment]
@GiaJordan Thank you for working on the cross-linking check. Nice progress! Feel free to ping me whenever you think the function is ready to try either this `feature-branch` or `develop` branch after you merge. We could work together to implement it into the DCA!

**Comment by rrchai** [comment]
@GiaJordan Great work.
I am not so familiar with the `schematic`. Would you mind telling me a bit more how to use this feature? 
- is there any other requirements, besides re-installing the `schematic`
- is the cross-linking part of `validateManifest`?

**Comment by GiaJordan** [comment]
@rrchai
Thank you!
Yes, this feature is implemented as a validation rule, so after reinstalling schematic you'll have to add the validation rule to whichever attribute in the data model you want it to apply to, convert the model again, and then validate the manifest. More information is available [here](https://sagebionetworks.jira.com/wiki/spaces/SCHEM/pages/2473623559/The+Data+Model+Schema#A.-Schema-properties-and-relationships), at the end of the section `A. Schema properties and relationships`

**Comment by rrchai** [comment]
@milen-sage What are essential cross-linking checks you think I can try to add for HTAN?

**Comment by milen-sage** [comment]
Yes, let's start with this one:
- ensure that in a manifest of type `scRNA-seq Level 1` the attribute (i.e. column) `HTAN Parent Biospecimen ID` matches at least one value of `HTAN Biospecimen ID` in manifests of type `Biospecimen`

@GiaJordan could you guide @rrchai in creating the validation rule? Since it's the first time we're testing this in a realistic schema it could be easier. 

In the future, the HTAN team can add these validation rules to the schema.

**Review by milen-sage — COMMENTED** [review]
@mialy-defelice and @GiaJordan to me the PR looks good! There is quite a bit of logic and some of it is described in configs. The latter is great for maintainability! However I can't reproduce and execute the resulting code logic in my brain. The best way to confirm the logic is working is testing the code. Once we have transitioned our prod systems to use the main branch and not develop, I'd be ok with merging this and similar PRs in develop. But for now we are still pulling from the develop branch in a few of our prod shiny apps. We can discuss at CR today on a good strategy to test #630 and this PR.

**Comment by milen-sage** [comment]
@GiaJordan the code logic looks good to me. How the code gets executed partially depends on config settings (e.g.  [great expectation config](https://github.com/Sage-Bionetworks/schematic/pull/614/files#r838808067)) and since I haven't played with the library I can't reproduce the expected behavior in my head :) 

Once we confirm with more testing downstream  (e.g. in DCA) that the refactor reproduces the past validation behavior and the additional new features work  as expected - we don't need to worry too much more about the configs and we'll be good to go. 

The code structure and docstrings are good otherwise!

**Comment by rrchai** [comment]
@GiaJordan I have tested this PR in DCA and I can make new validation rules work on the HTAN data, but there are three issues I had: 
1. after re-install the schematic, I have to re-install the markupsafe ` pip install markupsafe==2.0.1`, otherwise it cannot import name 'soft_unicode' from 'markupsafe'.
2. I have to comment out `ge_helpers.build_expectation_suite()` and reinstall the schematic, otherwise I will get below error when I validate using HTAN data model rules:
```
File "/home/rchai/ShinyApps/dc-test/.venv/lib/python3.8/site-packages/schematic/models/GE_Helpers.py", line 104, in build_expectation_suite
rule = self.sg.get_node_validation_rules(col)[0]
IndexError: list index out of range
```
3. I also have to run `great_expectations init` before app initiates.

I am not sure if all those ^ are needed fix or just me. Otherwise, the new validation from this PR can work on HTAN data.

**Comment by GiaJordan** [comment]
@rrchai 
> @GiaJordan I have tested this PR in DCA and I can make new validation rules work on the HTAN data, but there are three issues I had:
> 
> 1. I have to re-install the markupsafe ` pip install markupsafe==2.0.1`, otherwise it cannot import name 'soft_unicode' from 'markupsafe'.

We [merged a PR](https://github.com/Sage-Bionetworks/schematic/pull/636) into develop at code review yesterday that should handle this. I updated this branch to reflect those changes a little bit ago this morning. Can you pull the latest version of this branch and see if the version of markupsafe installed with poetry is enough?

> 2. I have to comment out `ge_helpers.build_expectation_suite()` in `validate_manifest.py` and reinstall the schematic, otherwise I will get below error when I validate using HTAN data model:
>    ```
>    File "/home/rchai/ShinyApps/dc-test/.venv/lib/python3.8/site-packages/schematic/models/GE_Helpers.py", line 104, in build_expectation_suite
>    rule = self.sg.get_node_validation_rules(col)[0]
>    IndexError: list index out of range
>    ```

I'll look into this error more to see why its coming up! Can you message me the `model.csv` and manifest you're using when you run into this?

> 3. I also have to run `great_expectations init` before app initiates.

Running `great_expectations init` is a normal step before use that I thought we could avoid by including the files generated from the init step. @milen-sage  I can update the documentation to reflect the necessity of this step since including the files wasn't enough.

**Comment by milen-sage** [comment]
@GiaJordan it sounds like resolving the conflicts on this PR along with ironing out the errors with @rrchai is what remains on this PR. @ychae we don't need to open an issue for that per @GiaJordan's [comment](https://github.com/Sage-Bionetworks/schematic/issues/577#issuecomment-1095628075), but let's prioritize for this "half" sprint.

**Comment by rrchai** [comment]
I have tested the new validation with @GiaJordan help. We figured out that `reticulate` in not happy with `great expectation`. Thus, it is good to add a para to enable/disable GE, if @GiaJordan has not implemented yet.

**Comment by rrchai** [comment]
@milen-sage @GiaJordan  
The current cross-manifests validation will download all manifests across all the projects. If we going to validate all submitted manifests in a project of HTAN, it will download hundreds of manifests in HTAN fileview each iteration (submitted manifest). Although, it should be less affected on users, it will take a lot of time as an admin for HTAN in DCA.

@milen-sage  Is cross-manifest rule required to also validate the manifest cross different projects? Take your example:
>ensure that in a manifest of type scRNA-seq Level 1 the attribute (i.e. column) HTAN Parent Biospecimen ID matches at least one value of HTAN Biospecimen ID in manifests of type Biospecimen

If we validate ^ rule on HTAN BU's `scRNA-seq Level 1` manifest, do we need to search for `HTAN Biospecimen ID` in `Biospecimen` manifests in other projects, e.g. Duke, CenterA? 

If no for ^, to speed up the validation, I highly recommend to 
1. add a params to subset a project
2. enable to provide an option to use array of file path to avoid duplicated downloadings. I believe it will be beneficial for iteration and current dashboard

**Comment by milen-sage** [comment]
> add a param to subset one or multiple project(s), or only download needed manifests

@rrchai yep - good point! @GiaJordan apologies for the confusion in communication here. Could we scope x-manifest validation rules to work with manifests only w/in a specified list of projects? In most cases this list would contain a single project, but we'd like to have flexibility and potentially do validation across multiple projects. 

@rrchai, we've been discussing adding annotations to manifests so that we can download only the ones we need for x-manifest validation. We'll see what's the performance with the project-scoping above and then can add this filtering feature for additional speed up as needed. 

> provide an option to use array of file path to avoid duplicated downloadings. I believe it will be beneficial for iteration and current dashboard

@GiaJordan we chatted about that^ - adding a parameter allowing passing filepaths to manifests and using these local manifest files would avoid downloading manifests if they have already been downloaded. @rrchai we can tackle that as well after adding the project-scoped validation above.

**Review by milen-sage — APPROVED** [review]
It looks good! 

After @GiaJordan merges this into develop, @ychae could you coordinate with stakeholders on deploying to a staging environment first, before using in production? Other than introducing a few features - e.g. new validation rules documented on schemahub - this PR also substantially changes how validation rules are processed behind the scenes. So it'd be great to do one more pass of testing in staging after merging to develop. The currently affected stakeholders are:
- HTAN - Rong can deploy to staging and Elvira and Adam can test the features
- NF - Robert can deploy to staging and NF team can test some of the features

Once we confirm this works in staging, @linglp let's do a schematic release following our dev cycle. @ychae can help coordinate when the release is ready to go.

**Comment by ychae** [comment]
@milen-sage added to both HTAN and NF stakeholder meeting agendas! I'm meeting with HTAN tomorrow and NF next Tuesday and should have a better idea of testing timelines after that. Please let me know if we need to inform the stakeholders sooner.

**Pull request merged**